core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java

```diff
@@ -0,0 +1,275 @@
+   * Add a new status to the cache.
+   *
+   * The last component of the path in the {@link AlluxioURI} must match the result of
+   * {@link UfsStatus#getName()}. This method overrides any status currently cached for the same
+   * URI.
+   *
+   * @param path the Alluxio path to key on
+   * @param status the ufs status to store
+   * @throws IllegalArgumentException if the status name doesn't match the final URI path component
+   */
+  public void addStatus(AlluxioURI path, UfsStatus status) {
+    UfsStatus prev = mStatuses.putIfAbsent(path, status);
+    if (!path.getName().equals(status.getName())) {
+      throw new IllegalArgumentException(
+          String.format("path name %s does not match ufs status name %s",
+              path.getName(), status.getName()));
+    }
+  }
+
+  /**
+   * Add a parent-child mapping to the status cache.
+   *
+   * All child statuses added via this method will be available via {@link #getStatus(AlluxioURI)}.
+   *
+   * @param path the directory inode path which contains the children
+   * @param children the children of the {@code path}
+   * @throws IllegalArgumentException when {@code path} already exists or if any child already
+   *                                  exists
+   */
+  public void addChildren(AlluxioURI path, Collection<UfsStatus> children) {
+    UfsStatus status = mStatuses.get(path);
+    // If this path doesn't yet exist, we can't keep track of the parent-child relationship
+    // We can still add statuses to the cache regardless
+    if (status != null) {
+      mChildren.computeIfAbsent(path, ufsStatus -> new ConcurrentHashSet<>()).addAll(children);
+    }
+    children.forEach(child -> {
+      AlluxioURI childPath = path.joinUnsafe(child.getName());
+      addStatus(childPath, child);
+    });
+  }
+
+  /**
+   * Remove a status from the cache.
+   *
+   *  Any children added to this status will remain in the cache.
+   *
+   * @param path the path corresponding to the {@link UfsStatus} to remove
+   * @return the removed UfsStatus
+   */
+  public UfsStatus remove(AlluxioURI path) {
+    UfsStatus removed = mStatuses.remove(path);
+    if (removed == null) {
+      return null;
+    }
+
+    mChildren.remove(path); // ok if there aren't any children
+    return removed;
+  }
+
+  /**
+   * Get the UfsStatus from a given AlluxioURI.
+   *
+   * @param path the path the retrieve
+   * @return The corresponding {@link UfsStatus} or {@code null} if there is none stored
+   */
+  public UfsStatus getStatus(AlluxioURI path) {
+    return mStatuses.get(path);
+  }
+
+  /**
+   * Fetches children of a given alluxio path, stores them in the cache, then returns them.
+   *
+   * Children can be returned in a few ways
+   * 1. Children already exist in the internal index. We simply return them
+   * 2. If children did not already exist in the index, then check if there was a scheduled
+   * prefetch job running for this path. If so, wait for the job to finish and return the result.
```

Comment:
What happens if there is no scheduled prefetch job? Is that what (3) is?

Reply:
`If no prefetch job` - isn't that what 3 says?